pkg/kv/kvserver/gc/gc.go

```diff
@@ -349,12 +321,37 @@ func processReplicatedKeyRange(
 					// All transactions in txnMap may be PENDING and
 					// cleanupIntentsFn will push them to finalize them.
 					info.PushTxn++
+
+					// Check if we reached max txn batch size that intent
+					// resolver can accept without splitting and push if
+					// needed.
+					if len(pendingTxns) >= maxTxnsPushPerBatch {
+						if err := cleanupIntentBatch(); err != nil {
```

Comment:
Could you try out the pattern with the helper struct I left on the last commit? I think that will allow you a) to test the batching in more isolation b) it de-noises the business logic here c) it avoids closing over vars from the outer scope in `cleanupIntentBatch` which is a common source of confusion and bugs.